CircuitVerse/CircuitVerse
# Testbench/Engine: Add Engine support for combinational and sequential tests

Changes and flow:
 - UI for creating tests (for now served as a static html file)
 - Implemented testbench engine for Combinational Tests
 - Validation for I/O, Clock and Reset pins
 - Runs tests as fast as possible without rendering(to increase speed) on Simulator
 - Currently uses Labels for identification of I/O. To be deprecated in favour of identifiers

 * Testbench engine stops the clock and runs tests by controlling the clock itself.
   This is required so that tests are not throttled by clock time.
 * Testbench engine returns test results with observed output next to expected output
 * Testbench engine returns with failure message on validation failure

Testing:
 - Create a test at `/testbench`. Clicking the `Save` button will log the test JSON in the console
 - Paste this JSON on Simulator under Tools > Use TestBench
 - Click `Run All`
 - Test Results are displayed in the console for now. Console logs an Array of groups
 - for each group, group.outputs has two fields - values (expected values) and result (observed values)

Test  Creator UI:
![Screenshot from 2021-06-17 12-38-54](https://user-images.githubusercontent.com/53974118/122348701-0f4aad00-cf69-11eb-92b6-b2036c3f9812.png)

TestBench Manual Mode UI
![Screenshot from 2021-06-22 17-21-52](https://user-images.githubusercontent.com/53974118/123053176-4285b400-d421-11eb-84cc-87cc28f0c55c.png)
